contracts/SafeProtocolMediator.sol

```diff
@@ -0,0 +1,241 @@
+            }
+        }
+
+        emit ActionsExecuted(address(safe), transaction.metaHash, transaction.nonce);
+    }
+
+    /**
+     * @notice This function executes a delegate call on a safe if the module is enabled and
+     *         root access it granted.
+     * @param safe A Safe instance
+     * @param rootAccess A struct of type SafeRootAccess containing information of about the action to be executed.
+     *                   Users can add logic to validate metahash through a transaction guard.
+     * @return data bytes types containing the result of the executed action.
+     */
+    function executeRootAccess(
+        ISafe safe,
+        SafeRootAccess calldata rootAccess
+    ) external override onlyEnabledModule(safe) returns (bytes memory data) {
+        SafeProtocolAction memory safeProtocolAction = rootAccess.action;
+
+        if (!ISafeProtocolModule(msg.sender).requiresRootAccess() || !enabledModules[address(safe)][msg.sender].rootAddressGranted) {
+            revert ModuleRequiresRootAccess(msg.sender);
+        }
+
+        bool success;
+        (success, data) = safe.execTransactionFromModuleReturnData(
+            safeProtocolAction.to,
+            safeProtocolAction.value,
+            safeProtocolAction.data,
+            1
+        );
+        if (success) {
+            emit RootAccessActionExecuted(address(safe), rootAccess.metaHash);
+        } else {
+            revert RootAccessActionExecutionFailed(address(safe), rootAccess.metaHash);
+        }
+    }
+
+    /**
+     * @notice Called by a Safe to enable a module on a Safe. To be called by a safe.
+     * @param module ISafeProtocolModule A module that has to be enabled
+     * @param allowRootAccess Bool indicating whether root access to be allowed.
+     */
+    function enableModule(address module, bool allowRootAccess) external noZeroOrSentinelModule(module) {
+        if (enabledModules[msg.sender][module].nextModulePointer != address(0)) {
+            revert ModuleAlreadyEnabled(msg.sender, module);
+        }
+
+        bool requiresRootAccess = ISafeProtocolModule(module).requiresRootAccess();
+        if (allowRootAccess != requiresRootAccess) {
+            revert ModuleAccessMismatch(module, requiresRootAccess, allowRootAccess);
+        }
+
+        if (enabledModules[msg.sender][SENTINEL_MODULES].nextModulePointer == address(0)) {
+            // The circular linked list has not been initialised yet for msg.sender. So, do it now.
+            enabledModules[msg.sender][SENTINEL_MODULES] = ModuleAccessInfo(false, SENTINEL_MODULES);
+        }
+
+        enabledModules[msg.sender][address(module)] = ModuleAccessInfo(
+            allowRootAccess,
+            enabledModules[msg.sender][SENTINEL_MODULES].nextModulePointer
+        );
+        enabledModules[msg.sender][SENTINEL_MODULES] = ModuleAccessInfo(false, address(module));
+
+        emit ModuleEnabled(msg.sender, address(module), allowRootAccess);
```

Comment:
`module` is already an address, why the conversion?

Reply:
In previous implementation it was`ISafeProtocolModule` then changed to address because code needs address more frequent than any interface method. I have removed the conversion.